BUILDING.md

```diff
@@ -0,0 +1,578 @@
+
+
+## The Tools
+
+`make` is where it all begins. With `make`, we build the cockroach binary, run unit tests, lint,
+build auxilary tools (roachprod, roachtest, workload, ...), etc.
+
+`build/builder.sh` lets us run `make` and other other tools within docker containers. Additionaly,
+it provides the ability to build and manage docker images.
+
+`build/builder/mkrelease.sh` is used to build release binaries of CockroachDB, including
+cross-compiling for different platforms. It must be run inside the `cockroachdb/builder` docker
+container (usually via `build/builder.sh`).
+
+The `build/teamcity-*.sh` scripts are how most TeamCity build configurations interact with the
+`cockroach` repository. These scripts use `build/builder.sh` to run `make` and
+`build/builder/mkrelease.sh` on TeamCity agents.
+
+The `build/packer/` directory contains what is needed for building VM images for the TeamCity
+agents.
+
+The `build/node-run.sh` script is a hack that pipes NodeJS's stdout and stderr through `cat` which
+prevents the actual stdout and stderr streams from being infected with non-blocking I/O. The
+comments in the script fully explain the scenario. It's used in the Makefile for each node-related
+command.
+
+The `build/upload-coverage.sh` script generates code coverage reports of all packages in the `pkg`
+directory and uploads the coverage profile to codecov.io. It's run via `make upload-coverage`.
+
+The `bin/uptodate` program is built from `pkg/cmd/uptodate` and used to check if c-deps are
+up-to-date or need to be rebuilt.
+
+The `build/werror.sh` script promotes warnings produced by a command to fatal errors, like GCC's
+`-Werror` flag.
+
+The `bin/execgen` program is built from `pkg/sql/colexec/execgen/cmd/execgen` and used to generate
+templated code related to columnarized execution.
+
+A slew of additional binaries deposited in `bin/` can be built using `make`. See [Go binary
+dependency file generation](#go-binary-dependency-file-generation) below for the list.
+
+
+# Tool-Specific Details
+
+## Make
+
+The CockroachDB makefile is large and complex. Here we focus on summarizing and clarifying it.
+
+`make help` gives an overview of available targets, variables and usage examples. We do not
+duplicate that information here.
+
+There are many detailed and helpful comments throughout the Makefile. When information from these
+comments is included in this document, it is only in summary.
+
+This make file can:
+
+- run on GNU and BSD systems (many commands have flags set by variables)
+- be run with different versions of Make (not just GNU make)
+- cross-compile cockroach to darwin, linux-gnu, linux-musl, and windows
+
+
+### Approximate Sections of the Makefile
+
+Functionality is mostly split into relevant sections. These are the approximate line numbers for
+each section (they will surely change). Additonal details on each section are included below.
+
+- Initialization: 0 - 374
```

Comment:
Hmm... there's no way this is going to stay up to date, right? I don't see this part being useful though I appreciate the effort it must have taken!